src/common/intrusive_lru.h

```diff
@@ -203,6 +216,21 @@ class intrusive_lru {
       }
   }
 
+  /// drop all elements from lru, invoke f on any with outstanding references
+  template <typename F>
+  void clear(F &&f) {
+    evict(0);
+    assert(unreferenced_list.empty());
+    for (auto &i: lru_set) {
+      std::invoke(f, static_cast<T&>(i));
+      i.lru = nullptr;
+      assert(i.is_invalidated());
+    }
+    lru_set.clear_and_dispose([](auto *i){
+      assert(i->use_count > 0); /* don't delete, still has a ref count */
```

Comment:
Except for asserting the use_cont, are we trying to erase the elements from the container without actually destructing them?
Should we assert that the lru_set is empty afterwards?

---

note: clear_and_dispose excepts a disposer function, however we aren't actually deleting the element.

```
   //! <b>Requires</b>: Disposer::operator()(pointer) shouldn't throw.
   //!
   //! <b>Effects</b>: Erases all the elements of the container.
   //!   No destructors are called.
   //!   Disposer::operator()(pointer) is called for the removed elements.
   //! 
   //! <b>Throws</b>: Nothing.
   //! 
   //! <b>Complexity</b>: Linear to the number of elements of the list.
   //! 
   //! <b>Note</b>: Invalidates the iterators to the erased elements.
   template <class Disposer>
   void clear_and_dispose(Disposer disposer)
```

Reply:
Yes, we are trying to clear the container without destroying the elements.

Reply:
nit: we can assert that lru_set is empty